File: node/src/worksheet.ts
import type { ResolvedConversionOptions } from "./index.js";
import type { SharedStringStore } from "./shared-strings.js";
import type { CellStyleStore } from "./styles.js";
import type { MarkdownWriter } from "./writer.js";
import type { XlsxArchive } from "./zip.js";
import { attribute, parseXml } from "./xml.js";

export interface WorksheetBounds {
  firstRow: number;
  lastRow: number;
  firstColumn: number;
  lastColumn: number;
}

interface Cell {
  row: number;
  column: number;
  type: string;
  style: number;
  value: string;
}

export async function scanWorksheet(
  archive: XlsxArchive,
  path: string,
  options: ResolvedConversionOptions,
): Promise<WorksheetBounds> {
  const bounds: WorksheetBounds = {
    firstRow: Number.MAX_SAFE_INTEGER,
    lastRow: 0,
    firstColumn: Number.MAX_SAFE_INTEGER,
    lastColumn: 0,
  };
  await visitCells(archive, path, options, (cell) => {
    if (cell.value.length === 0) return;
    bounds.firstRow = Math.min(bounds.firstRow, cell.row);
    bounds.lastRow = Math.max(bounds.lastRow, cell.row);
    bounds.firstColumn = Math.min(bounds.firstColumn, cell.column);
    bounds.lastColumn = Math.max(bounds.lastColumn, cell.column);
  });
  return bounds;
}

export async function writeWorksheet(
  archive: XlsxArchive,
  path: string,
  bounds: WorksheetBounds,
  strings: SharedStringStore,
  styles: CellStyleStore,
  writer: MarkdownWriter,
  options: ResolvedConversionOptions,
): Promise<void> {
  const values = new Map<number, string>();
  let currentRow = bounds.firstRow;
  let headerWritten = false;
  await visitCells(
    archive,
    path,
    options,
    (cell) => {
      if (cell.row < bounds.firstRow || cell.row > bounds.lastRow) return;
      while (cell.row > currentRow) {
        writeRow(writer, values, bounds.firstColumn, bounds.lastColumn);
        if (!headerWritten) {
          writeSeparator(writer, bounds.lastColumn - bounds.firstColumn + 1);
          headerWritten = true;
        }
        values.clear();
        currentRow++;
      }
      values.set(cell.column, formatValue(cell, strings, styles));
    },
    writer,
  );
  while (currentRow <= bounds.lastRow) {
    writeRow(writer, values, bounds.firstColumn, bounds.lastColumn);
    if (!headerWritten) writeSeparator(writer, bounds.lastColumn - bounds.firstColumn + 1);
    headerWritten = true;
    values.clear();
    currentRow++;
  }
}

async function visitCells(
  archive: XlsxArchive,
  path: string,
  options: ResolvedConversionOptions,
  visit: (cell: Cell) => void,
  writer?: MarkdownWriter,
): Promise<void> {
  await archive.withEntry(path, async (stream) => {
    let inferredRow = 0;
    let currentRow = 0;
    let inferredColumn = 0;
    let cell: Cell | undefined;
    let capture = false;
    await parseXml(stream, {
      open(tag) {
        if (tag.local === "row") {
          inferredRow++;
          currentRow = positiveInteger(attribute(tag, "r"), inferredRow);
          inferredRow = currentRow;
          inferredColumn = 0;
          if (currentRow > options.maximumRows) throw new Error("The worksheet exceeds the row limit.");
        } else if (tag.local === "c") {
          const reference = attribute(tag, "r");
          const column = reference ? parseColumn(reference) : inferredColumn + 1;
          inferredColumn = column;
          if (column > options.maximumColumns) throw new Error("The worksheet exceeds the column limit.");
          cell = {
            row: currentRow,
            column,
            type: attribute(tag, "t") ?? "",
            style: positiveInteger(attribute(tag, "s"), 0),
            value: "",
          };
        } else if (tag.local === "v" && cell) {
          capture = true;
        } else if (tag.local === "t" && cell?.type === "inlineStr") {
          capture = true;
        }
      },
      text(value) {
        if (capture && cell) cell.value += value;
      },
      close(tag) {
        if (tag.local === "v" || tag.local === "t") capture = false;
        else if (tag.local === "c" && cell) {
          visit(cell);
          cell = undefined;
        }
      },
      afterChunk: () => writer?.drain() ?? Promise.resolve(),
    });
  });
}

function formatValue(cell: Cell, strings: SharedStringStore, styles: CellStyleStore): string {
  if (cell.type === "s") return strings.get(cell.value);
  if (cell.type === "b") return cell.value === "1" ? "TRUE" : "FALSE";
  const number = Number(cell.value);
  if (cell.type === "" && cell.value.length > 0 && Number.isFinite(number)) {
    return styles.format(cell.style, number, cell.value);
  }
  return cell.value;
}

function writeRow(writer: MarkdownWriter, values: Map<number, string>, first: number, last: number): void {
  writer.write("|");
  for (let column = first; column <= last; column++) writer.write(` ${escapeCell(values.get(column) ?? "")} |`);
  writer.write("\r\n");
}

function writeSeparator(writer: MarkdownWriter, columns: number): void {
  writer.write(`|${" --- |".repeat(columns)}\r\n`);
}

function escapeCell(value: string): string {
  return value
    .replaceAll("\\", "\\\\")
    .replaceAll("|", "\\|")
    .replaceAll("\r\n", "<br>")
    .replaceAll("\r", "<br>")
    .replaceAll("\n", "<br>");
}

function positiveInteger(value: string | undefined, fallback: number): number {
  const parsed = Number(value);
  return Number.isInteger(parsed) && parsed > 0 ? parsed : fallback;
}

function parseColumn(reference: string): number {
  let column = 0;
  for (const character of reference) {
    if (character < "A" || character > "Z") break;
    column = column * 26 + character.charCodeAt(0) - 64;
  }
  if (column === 0 || !Number.isSafeInteger(column)) throw new Error("A cell reference is invalid.");
  return column;
}

File: rust/src/worksheet.rs
use std::collections::BTreeMap;
use std::io::{BufReader, Read, Write};

use quick_xml::Reader;
use quick_xml::events::Event;
use zip::read::ZipFile;

use crate::shared_strings::SharedStringStore;
use crate::styles::CellStyleStore;
use crate::xml::{attribute, local_name, text_value};
use crate::{ConversionOptions, ConversionResult, invalid_data};

pub(crate) struct WorksheetBounds {
    first_row: usize,
    last_row: usize,
    first_column: usize,
    last_column: usize,
}

impl WorksheetBounds {
    pub(crate) fn is_empty(&self) -> bool {
        self.last_row == 0
    }
}

pub(crate) fn scan_worksheet<R: Read>(
    entry: ZipFile<'_, R>,
    options: &ConversionOptions,
) -> ConversionResult<WorksheetBounds> {
    let mut bounds = WorksheetBounds {
        first_row: usize::MAX,
        last_row: 0,
        first_column: usize::MAX,
        last_column: 0,
    };
    visit_cells(entry, options, |row, column, _, _, value| {
        if !value.is_empty() {
            bounds.first_row = bounds.first_row.min(row);
            bounds.last_row = bounds.last_row.max(row);
            bounds.first_column = bounds.first_column.min(column);
            bounds.last_column = bounds.last_column.max(column);
        }
        Ok(())
    })?;
    Ok(bounds)
}

pub(crate) fn write_worksheet<R: Read, W: Write>(
    entry: ZipFile<'_, R>,
    bounds: &WorksheetBounds,
    strings: &mut SharedStringStore,
    styles: &CellStyleStore,
    output: &mut W,
    options: &ConversionOptions,
) -> ConversionResult<()> {
    let mut row_values = BTreeMap::new();
    let mut current_row = bounds.first_row;
    let mut header_written = false;
    visit_cells(entry, options, |row, column, cell_type, style, value| {
        if row < bounds.first_row || row > bounds.last_row {
            return Ok(());
        }
        while row > current_row {
            write_row(output, &row_values, bounds.first_column, bounds.last_column)?;
            if !header_written {
                write_separator(output, bounds.last_column - bounds.first_column + 1)?;
                header_written = true;
            }
            row_values.clear();
            current_row += 1;
        }
        row_values.insert(
            column,
            format_value(cell_type, style, value, strings, styles)?,
        );
        Ok(())
    })?;

    while current_row <= bounds.last_row {
        write_row(output, &row_values, bounds.first_column, bounds.last_column)?;
        if !header_written {
            write_separator(output, bounds.last_column - bounds.first_column + 1)?;
            header_written = true;
        }
        row_values.clear();
        current_row += 1;
    }
    Ok(())
}

fn visit_cells<R: Read, F>(
    entry: ZipFile<'_, R>,
    options: &ConversionOptions,
    mut visitor: F,
) -> ConversionResult<()>
where
    F: FnMut(usize, usize, &str, usize, &str) -> ConversionResult<()>,
{
    let mut reader = Reader::from_reader(BufReader::new(entry));
    let mut buffer = Vec::new();
    let mut inferred_row = 0;
    let mut current_row = 0;
    let mut inferred_column = 0;
    let mut current_cell: Option<(usize, String, usize, String)> = None;
    let mut capture_value = false;
    let mut capture_inline_text = false;

    loop {
        match reader.read_event_into(&mut buffer)? {
            Event::Start(element) => match local_name(element.name().as_ref()) {
                b"row" => {
                    inferred_row += 1;
                    current_row = positive_usize(attribute(&reader, &element, b"r")?, inferred_row);
                    inferred_row = current_row;
                    inferred_column = 0;
                    if current_row > options.maximum_rows {
                        return Err(invalid_data("The worksheet exceeds the row limit."));
                    }
                }
                b"c" => {
                    let column = match attribute(&reader, &element, b"r")? {
                        Some(reference) => parse_column(&reference)?,
                        None => inferred_column + 1,
                    };
                    inferred_column = column;
                    if column > options.maximum_columns {
                        return Err(invalid_data("The worksheet exceeds the column limit."));
                    }
                    let cell_type = attribute(&reader, &element, b"t")?.unwrap_or_default();
                    let style = positive_usize(attribute(&reader, &element, b"s")?, 0);
                    current_cell = Some((column, cell_type, style, String::new()));
                }
                b"v" if current_cell.is_some() => capture_value = true,
                b"t" if current_cell
                    .as_ref()
                    .is_some_and(|cell| cell.1 == "inlineStr") =>
                {
                    capture_inline_text = true;
                }
                _ => {}
            },
            Event::Text(text) if capture_value || capture_inline_text => {
                current_cell
                    .as_mut()
                    .unwrap()
                    .3
                    .push_str(&text_value(&text)?);
            }
            Event::CData(text) if capture_value || capture_inline_text => {
                current_cell.as_mut().unwrap().3.push_str(&text.decode()?);
            }
            Event::End(element) => match local_name(element.name().as_ref()) {
                b"v" => capture_value = false,
                b"t" => capture_inline_text = false,
                b"c" => {
                    if let Some((column, cell_type, style, value)) = current_cell.take() {
                        visitor(current_row, column, &cell_type, style, &value)?;
                    }
                }
                _ => {}
            },
            Event::Eof => break,
            _ => {}
        }
        buffer.clear();
    }
    Ok(())
}

fn format_value(
    cell_type: &str,
    style: usize,
    value: &str,
    strings: &mut SharedStringStore,
    styles: &CellStyleStore,
) -> ConversionResult<String> {
    match cell_type {
        "s" => strings.get(value),
        "b" => Ok(if value == "1" { "TRUE" } else { "FALSE" }.to_owned()),
        "" => match value.parse::<f64>() {
            Ok(number) => Ok(styles.format(style, number, value)),
            Err(_) => Ok(value.to_owned()),
        },
        _ => Ok(value.to_owned()),
    }
}

fn write_row<W: Write>(
    output: &mut W,
    values: &BTreeMap<usize, String>,
    first_column: usize,
    last_column: usize,
) -> ConversionResult<()> {
    write!(output, "|")?;
    for column in first_column..=last_column {
        write!(
            output,
            " {} |",
            escape_cell(values.get(&column).map(String::as_str).unwrap_or_default())
        )?;
    }
    output.write_all(b"\r\n")?;
    Ok(())
}

fn write_separator<W: Write>(output: &mut W, columns: usize) -> ConversionResult<()> {
    write!(output, "|")?;
    for _ in 0..columns {
        write!(output, " --- |")?;
    }
    output.write_all(b"\r\n")?;
    Ok(())
}

fn escape_cell(value: &str) -> String {
    value
        .replace('\\', "\\\\")
        .replace('|', "\\|")
        .replace("\r\n", "<br>")
        .replace(['\r', '\n'], "<br>")
}

fn positive_usize(text: Option<String>, fallback: usize) -> usize {
    text.and_then(|value| value.parse().ok())
        .filter(|value| *value > 0)
        .unwrap_or(fallback)
}

fn parse_column(reference: &str) -> ConversionResult<usize> {
    let mut column = 0_usize;
    for byte in reference
        .bytes()
        .take_while(|byte| byte.is_ascii_uppercase())
    {
        column = column
            .checked_mul(26)
            .and_then(|value| value.checked_add((byte - b'A' + 1) as usize))
            .ok_or_else(|| invalid_data("A cell reference is invalid."))?;
    }
    if column == 0 {
        return Err(invalid_data("A cell reference is invalid."));
    }
    Ok(column)
}
